### PR TITLE
Cleanup test containers after failed startup attempts

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

This allows using `container.withStartupAttempts()` together with the `ReportLeakedContainers` test execution listener.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java [hunk 1]
```diff
@@ -58,7 +62,40 @@ public static Closeable startOrReuse(GenericContainer<?> container)
         if (reuse) {
             return () -> {};
         }
-        return container::stop;
+        if (container.getStartupAttempts() <= 1) {
+            return container::stop;
+        }
+        return () -> {
+            container.stop();
+            // if there were multiple startup attempts, failed containers need to be manually removed
```
Comment (hashhar) [inline testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java:70]: if the containers actually terminate then why is the cleanup even needed?

But yes if the containers stick around (e.g. retry due to timeout) then I understand.

Can you please explain what I miss?

Comment (nineinchnick) [reply testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java:70]: They're stopped, but not removed. The resource leak detector doesn't allow that

Comment (hashhar) [reply testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java:70]: > Now I started to wonder if stopped containers should not be considered leaks? Ryuk would clean them up anyway

I tend to agree.

cc: @findepi

**Comment by nineinchnick** [comment]
Let's try https://github.com/trinodb/trino/pull/21280 instead of this.

File: testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java (hunk 1 above)
Comment (findepi) [reply testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java:70]: stopped containers are also a leak (they take up disk space, for example. Unless we  want to remain independent from ryuk for cleanup (which i personally want).

it's good that we have https://github.com/trinodb/trino/pull/21280 as a workaround
ngl, this PR wasn't very pretty. probably a cleaner fix could be done directly in TC. cc @wendigo

**Pull request closed without merging**